File: app/src/main/java/com/VegaSolutions/lpptransit/lppapi/responseobjects/Station.java
package com.VegaSolutions.lpptransit.lppapi.responseobjects;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.android.gms.maps.model.LatLng;
import com.google.gson.annotations.SerializedName;

import java.util.List;

public class Station implements Parcelable {

    private final String name;

    @SerializedName("int_id")
    private final int intId;

    private final double latitude;
    private final double longitude;

    @SerializedName("ref_id")
    private final String refId;

    @SerializedName("route_groups_on_station")
    private final List<String> routeGroupsOnStation;

    protected Station(Parcel in) {
        name = in.readString();
        intId = in.readInt();
        latitude = in.readDouble();
        longitude = in.readDouble();
        refId = in.readString();
        routeGroupsOnStation = in.createStringArrayList();
    }

    public static final Creator<Station> CREATOR = new Creator<Station>() {
        @Override
        public Station createFromParcel(Parcel in) {
            return new Station(in);
        }

        @Override
        public Station[] newArray(int size) {
            return new Station[size];
        }
    };

    public String getName() {
        return name;
    }

    public int getIntId() {
        return intId;
    }

    public double getLatitude() {
        return latitude;
    }

    public double getLongitude() {
        return longitude;
    }

    public String getRefId() {
        return refId;
    }

    public List<String> getRouteGroupsOnStation() {
        return routeGroupsOnStation;
    }

    public LatLng getLatLng() {
        return new LatLng(latitude, longitude);
    }

    public boolean isCenter() {
        return Integer.valueOf(refId) % 2 != 0;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(name);
        dest.writeInt(intId);
        dest.writeDouble(latitude);
        dest.writeDouble(longitude);
        dest.writeString(refId);
        dest.writeStringList(routeGroupsOnStation);

    }

    @Override
    public String toString() {
        return "Station{" +
                "name='" + name + '\'' +
                ", int_id=" + intId +
                ", latitude=" + latitude +
                ", longitude=" + longitude +
                ", ref_id='" + refId + '\'' +
                ", route_groups_on_station=" + routeGroupsOnStation +
                '}';
    }
}

File: app/src/main/java/com/VegaSolutions/lpptransit/data/lppapi/data/Station.kt
package com.VegaSolutions.lpptransit.data.lppapi.data

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import com.google.android.gms.maps.model.LatLng
import com.google.gson.annotations.SerializedName

data class Station(
    @SerializedName("name")
    val name: String,

    @SerializedName("int_id")
    val intId: Int,

    @SerializedName("latitude")
    val latitude: Double,

    @SerializedName("longitude")
    val longitude: Double,

    @SerializedName("ref_id")
    val refId: String,

    @SerializedName("route_groups_on_station")
    val routeGroupsOnStation: List<String>,
) : Parcelable {

    val latLng: LatLng
        get() = LatLng(latitude, longitude)

    val towards: Boolean
        get() = refId.toIntOrNull()?.mod(2)?.equals(0) ?: false

    constructor(parcel: Parcel) : this(
        parcel.readString()!!,
        parcel.readInt(),
        parcel.readDouble(),
        parcel.readDouble(),
        parcel.readString()!!,
        parcel.createStringArrayList()!!
    ) {
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(name)
        dest.writeInt(intId)
        dest.writeDouble(latitude)
        dest.writeDouble(longitude)
        dest.writeString(refId)
        dest.writeStringList(routeGroupsOnStation)
    }

    companion object CREATOR : Creator<Station> {
        override fun createFromParcel(parcel: Parcel): Station {
            return Station(parcel)
        }

        override fun newArray(size: Int): Array<Station?> {
            return arrayOfNulls(size)
        }
    }
}
